routing/index_graph.cpp

```diff
@@ -150,6 +195,134 @@ void IndexGraph::GetNeighboringEdges(Segment const & from, RoadPoint const & rp,
   }
 }
 
+void IndexGraph::GetSegmentCandidateForJoint(Segment const & parent, bool isOutgoing,
+                                             vector<Segment> & childs)
+{
+  RoadPoint const roadPoint = parent.GetRoadPoint(isOutgoing);
+  Joint::Id const jointId = m_roadIndex.GetJointId(roadPoint);
+
+  if (jointId == Joint::kInvalidId)
+    return;
+
+  m_jointIndex.ForEachPoint(jointId, [&](RoadPoint const & rp)
+  {
+    RoadGeometry const & road = m_geometry->GetRoad(rp.GetFeatureId());
+    if (!road.IsValid())
+      return;
+
+    bool const bidirectional = !road.IsOneWay();
+    auto const pointId = rp.GetPointId();
+
+    if ((isOutgoing || bidirectional) && pointId + 1 < road.GetPointsCount())
+      childs.emplace_back(parent.GetMwmId(), rp.GetFeatureId(), pointId, isOutgoing);
+
+    if ((!isOutgoing || bidirectional) && pointId > 0)
+      childs.emplace_back(parent.GetMwmId(), rp.GetFeatureId(), pointId - 1, !isOutgoing);
+  });
+}
+
+void IndexGraph::ReconstructJointSegment(Segment const & parent,
+                                         vector<Segment> const & firstChilds,
+                                         vector<uint32_t> const & lastPointIds,
+                                         bool isOutgoing,
+                                         vector<JointEdge> & jointEdges,
+                                         vector<RouteWeight> & parentWeights)
+{
+  auto const step = [this](Segment const & from, Segment const & to, bool isOutgoing, SegmentEdge & edge)
+  {
+    if (IsRestricted(m_restrictions, from, to, isOutgoing))
+      return false;
+
+    RouteWeight weight = CalcSegmentWeight(isOutgoing ? to : from) +
+                         GetPenalties(isOutgoing ? from : to, isOutgoing ? to : from);
+    edge = SegmentEdge(to, weight);
+    return true;
+  };
+
+  for (size_t i = 0; i < firstChilds.size(); ++i)
+  {
+    auto const & firstChild = firstChilds[i];
+    auto lastPointId = lastPointIds[i];
+
+    uint32_t currentPointId = firstChild.GetPointId(!isOutgoing /* front */);
+    CHECK_NOT_EQUAL(currentPointId, lastPointId,
+                    ("Invariant violated, can not build JointSegment,"
+                     "started and ended in the same point."));
```

Comment:
Может ли тут оказаться самопересекающаяся фича? Такая фича может образовать JointSegment, который начинается и заканчивается в одной и той же точке. В osm такие фичи есть.
Обрати внимание на комменты к этому коду нашего проекта.
```
/// \brief This method should be called for an internal junction of the route with corresponding
/// |ingoingEdges|, |outgoingEdges|, |ingoingRouteEdge| and |outgoingRouteEdge|.
/// \returns false if the junction is an internal point of feature segment and can be considered as
/// a part of LoadedPathSegment and returns true if the junction should be considered as a beginning
/// of a new LoadedPathSegment.
bool IsJoint(IRoadGraph::TEdgeVector const & ingoingEdges,
             IRoadGraph::TEdgeVector const & outgoingEdges,
             Edge const & ingoingRouteEdge,
             Edge const & outgoingRouteEdge,
             bool isCurrJunctionFinish,
             bool isInEdgeReal)
{
  // When feature id is changed at a junction this junction should be considered as a joint.
  //
  // If a feature id is not changed at a junction but the junction has some ingoing or outgoing edges with
  // different feature ids, the junction should be considered as a joint.
  //
  // If a feature id is not changed at a junction and all ingoing and outgoing edges of the junction has
  // the same feature id, the junction still may be considered as a joint.
  // It happens in case of self intersected features. For example:
  //            *--Seg3--*
  //            |        |
  //          Seg4      Seg2
  //            |        |
  //   *--Seg0--*--Seg1--*
  // The common point of segments 0, 1 and 4 should be considered as a joint.
  if (!isInEdgeReal)
    return true;
...
```

Reply:
Похоже, что этот чек может сработать если волна накроет место с самопересекающейся фичей.

Reply:
как это разные points, с точки зрения номера, то `CHECK`, при правильной работе алгоритма не сработает